src/RequestCompressionMiddleware.php

```diff
@@ -0,0 +1,160 @@
+            && is_int($command['@request_min_compression_size_bytes'])
+            && $this->isValidCompressionSize($command['@request_min_compression_size_bytes'])
+        ) {
+            $this->minimumCompressionSize = $command['@request_min_compression_size_bytes'];
+        }
+        $nextHandler = $this->nextHandler;
+        $operation = $this->api->getOperation($command->getName());
+        $requestBodySize = $request->getBody()->getSize();
+        $compressionTrait = isset($operation['requestcompression'])
+            ? $operation['requestcompression']
+            : null;
+
+        if (!$this->shouldCompressRequestBody(
+            $compressionTrait,
+            $command,
+            $operation,
+            $requestBodySize
+        )) {
+            return $nextHandler($command, $request);
+        }
+
+        $this->encodings = $compressionTrait['encodings'];
+        $request = $this->compressRequestBody($request);
+
+        return $nextHandler($command, $request);
+    }
+
+    private function compressRequestBody(
+        RequestInterface $request
+    ) {
+        $fn = $this->determineEncoding();
+        if (is_null($fn)) {
+            return $request;
+        }
+
+        $body = $request->getBody()->getContents();
+        $compressedBody = $fn($body);
+
+        return $request->withBody(Psr7\Utils::streamFor($compressedBody))
+            ->withHeader('content-encoding', $this->encoding);
+    }
+
+    private function determineEncoding()
+    {
+        foreach ($this->encodings as $encoding) {
+            if (isset($this->encodingMap[$encoding])) {
+                $this->encoding = $encoding;
+                return $this->encodingMap[$encoding];
+            }
+        }
+        return null;
+    }
+
+    private function shouldCompressRequestBody(
+        $compressionTrait,
+        $command,
+        $operation,
+        $requestBodySize
+    ){
+        if ($compressionTrait) {
+            if (isset($command['@disable_request_compression'])
+                && $command['@disable_request_compression'] === true
+            ) {
+                return false;
+            } elseif ($this->hasStreamingTraitWithoutRequiresLength($command, $operation)
+                || $requestBodySize >= $this->minimumCompressionSize
+            ) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private function hasStreamingTraitWithoutRequiresLength($command, $operation)
+    {
+        foreach ($operation->getInput()->getMembers() as $name => $member) {
+            if (isset($command[$name])
+                && !empty($member['streaming'])
+                && empty($member['requiresLength'])
+            ){
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private function determineMinimumCompressionSize($config) {
+        if (isset($config['request_min_compression_size_bytes'])) {
```

Comment:
What gets returned when this isn't set?  I think this check is not necessary since it's always set in `_apply_request_min_compression_size_bytes`